src/artifact_providers/github.ts

```diff
@@ -81,46 +82,74 @@ export class GithubArtifactProvider extends BaseArtifactProvider {
    */
   private async listArtifact(
     revision: string,
-    page = 0
+    revisionDate?: string,
+    page = 0,
+    tries = 0,
   ): Promise<ArtifactItem> {
-    const { repoName, repoOwner } = this.config;
+    const { repoName: repo, repoOwner: owner } = this.config;
     const per_page = 100;
 
     // https://docs.github.com/en/free-pro-team@latest/rest/reference/actions#artifacts
     const artifactResponse = ((
       await this.github.request('GET /repos/{owner}/{repo}/actions/artifacts', {
-        owner: repoOwner,
-        repo: repoName,
+        owner: owner,
+        repo: repo,
         per_page,
         page,
       })
     ).data as unknown) as ArtifactList;
 
-    if (artifactResponse.total_count === 0) {
-      throw new Error(`Failed to discover any artifacts`);
-    }
+    const allArtifacts = artifactResponse.artifacts;
 
     // We need to find the archive where name maches the revision
-    const foundArtifacts = _.filter(
-      artifactResponse.artifacts,
+    const foundArtifacts = allArtifacts.filter(
       artifact => artifact.name === revision
     );
 
-    if (foundArtifacts.length === 0) {
-      if (artifactResponse.total_count > per_page * (page + 1)) {
-        return this.listArtifact(revision, page + 1);
-      }
-      throw new Error(`Can't find artifacts for revision \`${revision}\``);
+    if (foundArtifacts.length === 1) {
+      return foundArtifacts[0];
     }
 
     if (foundArtifacts.length > 1) {
       throw new Error(
         `Found multiple artifacts with the same revision \`${revision}\`\n` +
-          `Please make sure you job only uploads on set of artifacts.`
+          `Please make sure your job only uploads one set of artifacts.`
       );
     }
 
-    return foundArtifacts[0];
+    let checkNextPage = false;
+    if (artifactResponse.total_count > per_page * (page + 1)) {
+
+      if (revisionDate === undefined) {
+        revisionDate = (await this.github.git.getCommit({
+          owner,
+          repo,
+          commit_sha: revision
+        })).data.committer.date;
+      }
+      const lastArtifact = allArtifacts[allArtifacts.length-1];
+      checkNextPage = lastArtifact.created_at >= revisionDate;
```

Comment:
Is the "sorted by creation date" guarantee actually provided by the GitHub API we use?

Reply:
They don't document any order but it looks like it is inverse creation order from this: https://api.github.com/repos/getsentry/craft/actions/artifacts

And honestly, if it was random, that would be extremely inefficient for GitHub themselves too so I think this logic is safe.

I also checked to see if they support this on their v4 GraphQL API so we can enforce the ordering ourselves, but not there yet ☹️

Reply:
If enough data supports it, I guess we can assume that, but then please add a comment to the code somewhere that it's just our assumption.

Otherwise looks ok, can approve when it's tested.